Documentation/MyFirstContribution

```diff
@@ -0,0 +1,584 @@
+the Developer's Certificate of Origin 1.1 (see the SubmittingPatches [[dco]]
+header). If you wish to add some context to your change, go ahead with
+`git commit --amend`.
+
+=== Implementation
+
+It's probably useful to do at least something besides print out a string. Let's
+start by having a look at everything we get.
+
+Modify your `cmd_psuh` implementation to dump the args you're passed:
+
+----
+	printf("Your args (there are %i):\n", argc);
+	for (int i = 0; i < argc; i++) {
+		printf("%s\n", argv[i]);
+	}
+	printf("Your prefix:\n%s\n", prefix);
+----
+
+As you may expect, there's pretty much just whatever we give on the command
+line, including the name of our command. That's not so helpful. So what other
+context can we get?
+
+Add a line to `#include "config.h"`. Then, add the following bits:
+
+----
+const char *cfg_name;
+
+...
+
+git_config(git_default_config, NULL)
+if (git_config_get_string_const("user.name", &cfg_name) > 0)
+{
+	printf("No name is found in config\n");
+}
+else
+{
+	printf("Your name: %s\n", cfg_name);
+}
+----
+
+git_config(...) will grab the configuration from config files known to Git and
+apply standard precedence rules. git_config_get_string_const(...) will look up
+a specific key ("user.name") and give you the value. There are a number of
+single-key lookup methods like this one; you can see them all (and more info
+about how to use git_config()) in `Documentation/technical/api-config.txt`.
+
+You should see that the name printed matches the one you see when you run:
+
+----
+git config --get user.name
+----
+
+Great! Now we know how to check for values in the git config. Let's commit this
+too, so we don't lose our progress.
+
+----
+git add builtin/psuh.c
+git commit -sm "psuh: show parameters & config opts"
+----
+
+Still, it'd be nice to know what the user's working context is like. Let's see
+if we can print the name of the user's current branch. We can cheat off of the
+`git status` implementation; the printer is located in `wt-status.c` and we can
+see that the branch is held in a `struct wt_status`.  `wt_status_print()` gets
+invoked by `cmd_status()` in `builtin/commit.c`. Looking at that implementation
+we see the status config being populated like so:
+
+----
+status_init_config(&s, git_status_config);
+----
+
+But as we drill down, we can find that `status_init_config()` wraps a call
+to `git_config()`. Let's modify the code we wrote in the previous commit.
+
+----
+#include "wt-status.h"
+
+...
+
+// Add a wt_status to fill at the top.
+struct wt_status status;
+
+...
+
+// modify the prior code:
+wt_status_prepare(the_repository, &status);
+git_config(git_default_config, &status);
+
+...
+
+printf("Your current branch: %s\n", status.branch);
+----
+
+Run it again. Check it out - here's the (verbose) name of your current branch!
+
+Let's commit this as well.
+
+----
+git commit -sm "psuh: print the current branch"
+----
+
+TODO: ref & object read
```

Comment:
You mean, like, reading the `HEAD`? Good idea. The `dwim_ref("HEAD", 4, &oid, &branch_name)` and `oid_to_hex(&oid)` calls will come in handy ;-)

Reply:
I'm planning on working on this in parallel but sending the tutorial so far for RFC.